### PR TITLE
Support for developer comments

### LINKED ISSUE #115: Support dev comments
**Is your feature request related to a particular use-case?**

Dev comments could contain spelling mistakes too.

**Describe the solution you'd like to implement/see implemented**

A file containing `//` or `/*! .. */` based comments should also be spell checked.

**Describe alternatives you've considered**

Status quo. Do not check them.

### PR DESCRIPTION
**Pull request by snasphysicist**
This PR adds support for spellchecking developer comments in addition to documentation comments.

 * 🦚 Feature
snasphysicist
Closes #115

## Changes proposed by this PR:

Adds a new module which is responsible for extracting developer comments from the source, converting them into a format (`LiteralSet`) required by the spellchecker and adds them to the data structure which stores the content to be spellchecked.

This also adds a new dependency - `ra_ap_syntax`.

## Notes to reviewer:

The bulk of the code is in its own, new module (`developer.rs`). We can't use anything from `proc_macro2` since it throws away all information about developer comments so we need to start again from the source file/text, and the new code with tests is quite long (a few hundred lines).

The source is split into tokens, which are then passed through a series of transformations adding more information (location in source, then line and column at which they start, then "type" calculated from a regex). Only block and non-documentation line comments are retained and these are separately converted into `TrimmedLiteral`s/`LiteralSet`s.

The main entry point `extract_developer_comments` is called from `cluster.rs` and the created `LiteralSet`s are added to the cluster's vector of such sets. Since two parsers are currently in use (`proc_macro2` & `ra_ap_syntax`) first one is applied then the other. This means that the comments are initially added to the vector of sets not in the order in which they occur in the file. hence a new method is added to sort this vector by the first line on which each `LiteralSet` occurs.

## Specific TODOs

 * [X] Re-enable strict code/lint checks
 * [X] Lazy static for regexes
 * [X] ~~Fix one unexpectedly failing test for line comments - not sure about the meaning of 'adjacent' in `LiteralSet`~~ (turns out I was being an idiot!)
 * [X] Add parsed `LiteralSet`s to `CheckableChunk` 
 * [X] Document all functions, structs
 * [X] ~~Possibly replace regex with fancy_regex~~? (regex dependency was introduced by another change)
 * [X] Does all the new code live in the right place?
 * [X] Check/squash/tidy old commits

## 📜 Checklist

 * [X] Works on the `./demo` sub directory
 * [X] Test coverage is excellent and passes (IMO :) )
 * [X] Documentation is thorough (IMO :) )

### CODE REVIEW

File: src/documentation/literal.rs [hunk 1]
```diff
@@ -280,6 +283,34 @@ impl TryFrom<(&str, proc_macro2::Literal)> for TrimmedLiteral {
     }
 }
 
+impl TrimmedLiteral {
```
[the remark this answers is not in the document]
Comment (snasphysicist) [reply src/documentation/literal.rs:284]: Agreed - I also had the idea that using both parsing packages is a bit redundant now, but I didn't want to exceed the scope of #130 . I do think the trimmed literal works quite well though, so I wouldn't necessarily throw it out entirely!

File: src/documentation/developer.rs [hunk 2]
```diff
@@ -0,0 +1,747 @@
...
+      let literal = all_literals.get(1).unwrap();
+      assert!(literal.as_str().contains(content_2));
+    }
+  }
+
+  #[test]
+  fn test_non_adjacent_line_comments_put_in_different_literal_sets() {
+    let content_1 = " line comment 1 ";
+    let content_2 = " line comment 2 ";
+    let source = format!("//{}\nfn(){{}}\n//{}", content_1, content_2);
```
[the remark this answers is not in the document]
Comment (snasphysicist) [reply src/documentation/developer.rs:777]: Sure, I'll have a think about how/the best way to do that.

File: src/documentation/developer.rs [hunk 3]
```diff
@@ -0,0 +1,747 @@
...
+      assert!(literal.as_str().contains(content_1));
+    }
+    {
+      let literal_set = literal_sets.get(1).unwrap();
+      let all_literals = literal_set.literals();
+      assert_eq!(all_literals.len(), 1);
+      let literal = all_literals.get(0).unwrap();
+      assert!(literal.as_str().contains(content_2));
+    }
+  }
```
[the remark this answers is not in the document]
Comment (snasphysicist) [reply src/documentation/developer.rs:798]: Thanks! I was worried it was a bit too verbose.

File: src/documentation/developer.rs [hunk 4]
```diff
@@ -0,0 +1,747 @@
...
+  tokens_in.into_iter().map(|t| identify_token_type(t)).collect()
+}
+
+fn retain_only_developer_comments(tokens: Vec<TokenWithType>) -> Vec<TokenWithType> {
+  tokens.into_iter()
+      .filter(|t| t.kind != TokenType::Other)
+      .collect()
+}
+
+fn literal_set_from_block_comment(token: TokenWithType) -> Option<LiteralSet> {
```
[the remark this answers is not in the document]
Comment (snasphysicist) [reply src/documentation/developer.rs]: That makes sense - implemented

File: src/documentation/developer.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (snasphysicist) [reply src/documentation/developer.rs:781]: In the end, I felt `macro_rules` was too complex a solution for this problem and the test setup in many places could be simplified with a few helper functions. Let me know if there are any remaining bits you feel would be improved by further helper functions or macros.

**Comment by snasphysicist** [comment]
Hmm the CI is failing at compile time due to a problem in `ra_ap_parser`

```
error[E0658]: `while` is not allowed in a `const fn`
  --> /tmp/build/b4e0c68a/git-pull-request-resource/../cargo/registry/src/github.com-1ecc6299db9ec823/ra_ap_parser-0.0.23/src/token_set.rs:15:9
   |
15 | /         while i < kinds.len() {
16 | |             res |= mask(kinds[i]);
17 | |             i += 1
18 | |         }
   | |_________^
   |
   = note: see issue #52000 <https://github.com/rust-lang/rust/issues/52000> for more information
```

This doesn't occur locally when building with `rustc` 1.47.0, CI is using `rustc` 1.45.2 - not sure if that's the root of the issue.

**Comment by snasphysicist** [comment]
Hi @drahnr , I think this has been tidied to a point now where it's ready for a full review. I've tried to keep the commits relatively small and self-contained, but do let me know if you think it could benefit from some more squashing. Thanks!

**Comment by drahnr** [comment]
> Hmm the CI is failing at compile time due to a problem in `ra_ap_parser`
> 
> ```
> error[E0658]: `while` is not allowed in a `const fn`
>   --> /tmp/build/b4e0c68a/git-pull-request-resource/../cargo/registry/src/github.com-1ecc6299db9ec823/ra_ap_parser-0.0.23/src/token_set.rs:15:9
>    |
> 15 | /         while i < kinds.len() {
> 16 | |             res |= mask(kinds[i]);
> 17 | |             i += 1
> 18 | |         }
>    | |_________^
>    |
>    = note: see issue #52000 <https://github.com/rust-lang/rust/issues/52000> for more information
> ```
> 
> This doesn't occur locally when building with `rustc` 1.47.0, CI is using `rustc` 1.45.2 - not sure if that's the root of the issue.

`while let` in `const fn` was introduced in `1.46` https://blog.rust-lang.org/2020/08/27/Rust-1.46.0.html - I wanted to update the CI image but that doesn't work right now due to https://github.com/concourse/concourse/pull/6348 which just got merged last night.

I'll give it a whirl locally and update the CI image manually to get this going.

File: src/documentation/developer.rs [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+  pub line: usize,
+  /// The column where the first character of this token appears in the source file (0 indexed)
+  pub column: usize,
+}
+
+impl TokenWithType {
+  /// Convert a `TokenWithLineColumn` to a `TokenWithType`. The kind is worked out from the content
+  /// by checking against the developer block comment & line comment regexes.
+  fn from (token: TokenWithLineColumn) -> Self {
+    let kind = {
```
Comment (drahnr) [inline src/documentation/developer.rs:123]: nit: The outermost `{ .. };` are not necessary.

File: src/documentation/developer.rs [hunk 6]
```diff
@@ -0,0 +1,803 @@
...
+      }
+    };
+    Self { kind, content: token.content, line: token.line, column: token.column }
+  }
+}
+
+/// A convenience method that runs the complete 'pipeline' from string `source` file to all
+/// `LiteralSet`s that can be created from developer comments in the source
+pub fn extract_developer_comments(source: &str) -> Vec<LiteralSet> {
+  let tokens = retain_only_developer_comments(
```
Comment (drahnr) [inline src/documentation/developer.rs:139]: nit: I find it helpful to have a few local variables to be inspected upon debugging.

**Review by drahnr — COMMENTED** [review]
I have yet to test and update the CI image to rust 1.46.0 - but this looks very good!

**Comment by drahnr** [comment]
Adding one more dummy commit should re-trigger the CI which is now up to date regarding the rustc version.

**Comment by drahnr** [comment]
```
  13   │ /*
  14   │  * Also check thiz one
  15   │  */
  ```
  
  results in 
  
  ```
  error: spellcheck(Hunspell)
   --> /media/supersonic1t/projects/cargo-spellcheck/demo/src/nested/fragments.rs:14
    |
 14 |  * Also check thiz one
    |               ^^^^
    | - this, thin, whiz, or thigh
    |
    |   Possible spelling mistake found.
```

where I would like to detect if all of them start with ` * ` sequence and iff, elimnate that from the view, but I guess that's a stretch goal that can be done in a follow up PR.

Very nice work! Much appreciated!

File: src/documentation/developer.rs (hunk 5 above)
Comment (snasphysicist) [reply src/documentation/developer.rs]: Removed in latest push. Matter of taste, personally I find it a bit easier to ready with the extra brackets.

File: src/documentation/developer.rs (hunk 6 above)
Comment (snasphysicist) [reply src/documentation/developer.rs]: Good point, I've added a few intermediate variables in the functions that run pipelines of other functions in the latest push.

**Comment by snasphysicist** [comment]
@drahnr sorry I've been a bit slow to get these changes addressed. Thanks for getting the CI sorted!

I wasn't sure about the `* ` because technically it's just a common stylistic choice rather than something required by the language. Shouldn't be too hard to implement though. I'm happy to do this in the new year (or would make another good first time contributor issue).

**Review by drahnr — APPROVED** [review]
Looks very good!

Merry Christmas to you too! ⛄🎄

**Comment by drahnr** [comment]
Regarding ` * ` I'd be happy if you'd take on that, there are still plenty of low hanging fruits in the code base for starters.

Thanks again!

**Pull request merged**